src/coreclr/nativeaot/Runtime/unix/PalCreateDump.cpp

```diff
@@ -0,0 +1,572 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <stdint.h>
+#include <errno.h>
+#include <cwchar>
+#include <sal.h>
+#include "config.h"
+#include <pthread.h>
+#include <string.h>
+#include <ctype.h>
+#include <cstdarg>
+#include <unistd.h>
+#define __STDC_FORMAT_MACROS
+#include <inttypes.h>
+#include <sys/mman.h>
+#include <sys/types.h>
+#include <sys/stat.h>
+#include <signal.h>
+#if HAVE_PRCTL_H
+#include <sys/prctl.h>
+#include <sys/syscall.h>
+#endif
+#include <sys/wait.h>
+#include <sys/time.h>
+#include <sys/resource.h>
+#include <dlfcn.h>
+
+#ifdef __APPLE__
+#include <libproc.h>
+#include <sys/sysctl.h>
+#include <sys/posix_sem.h>
+#include <mach/task.h>
+#endif
+
+#ifdef __NetBSD__
+#include <sys/cdefs.h>
+#include <sys/param.h>
+#include <sys/sysctl.h>
+#include <kvm.h>
+#endif
+
+#ifdef __FreeBSD__
+#include <sys/sysctl.h>
+#include <sys/user.h>
+#endif
+
+#define _ASSERTE(x)
+
+#include <clrconfignocache.h>
+#include <generatedumpflags.h>
+
+// Crash dump generating program arguments. MAX_ARGV_ENTRIES is the max number
+// of entries if every createdump option/argument is passed.
+#define MAX_ARGV_ENTRIES 32 
+const char* g_argvCreateDump[MAX_ARGV_ENTRIES] = { nullptr };
+char* g_szCreateDumpPath = nullptr;
+char* g_ppidarg  = nullptr;
+
+/*++
+Function:
+    PlatformGetCurrentThreadId
+
+    Returns the current thread id
+*/
+
+#if defined(__linux__)
+#define PlatformGetCurrentThreadId() (uint32_t)syscall(SYS_gettid)
+#elif defined(__APPLE__)
+inline uint32_t PlatformGetCurrentThreadId() {
+    uint64_t tid;
+    pthread_threadid_np(pthread_self(), &tid);
+    return (SIZE_T)tid;
+}
+#elif defined(__FreeBSD__)
+#include <pthread_np.h>
+#define PlatformGetCurrentThreadId() (uint32_t)pthread_getthreadid_np()
+#elif defined(__NetBSD__)
+#include <lwp.h>
+#define PlatformGetCurrentThreadId() (uint32_t)_lwp_self()
+#else
+#define PlatformGetCurrentThreadId() (uint32_t)pthread_self()
+#endif
+
+/*++
+Function:
+    FormatInt
+
+    Helper function to format an uint32 as a string.
+
+--*/
+static
+char*
+FormatInt(uint32_t value)
+{
+    char* buffer = (char*)malloc(128);
```

Comment:
I think it would be easier to pass in a buffer and fill it. We can remove the entire allocation mechanism and do everything on the stack.

Reply:
Looks like we are also parsing 32/64 unsigned integers. We have constants in the runtime for the maximum length of these values. See below.

https://github.com/dotnet/runtime/blob/31bcc752fed783227eba08ece7e07a690aa12d11/src/coreclr/inc/utilcode.h#L273-L275

Reply:
The code saves away the values in the g_argvCreateDump/g_ppidarg globals so the malloc's are needed. 

Are those constants available to native aot code?  I do know that src/coreclr/inc is not in the search path.

Reply:
> The code saves away the values in the g_argvCreateDump/g_ppidarg globals so the malloc's are needed.

So it seems like we can create a static alloc for `g_ppidarg`

```cpp
char g_ppidarg[CONSTANT];
```

The `g_argvCreateDump` is fair, but we can also do the allocation outside of this function and pass it in. I was noting the clean-up logic in `PalCreateCrashDumpIfEnabled` which could be removed.

Reply:
Those defines don't seem to include the terminating \0 unless I'm reading it wrong somehow.  If they are used for an allocation and the number formatted is the max, won't it overflow the buffer?

Reply:
> Those defines don't seem to include the terminating \0 unless I'm reading it wrong somehow.

You're reading them correctly. Current use cases are for embedding in other strings so the null count was dropped. For this use case, you are right and should add 1 for null.